### PR TITLE
Add support for DPS symmetric key attestation

### PR DESCRIPTION
**Pull request by mrohera**
Summary: The changes here implement the client code needed to register with DPS in order for DPS to resolve the Edge device to the configured IoT Hub. More about this capability in DPS can be learned here: [https://docs.microsoft.com/en-us/azure/iot-dps/quick-create-simulated-device-symm-key](https://docs.microsoft.com/en-us/azure/iot-dps/quick-create-simulated-device-symm-key)

Overall the changes deal with parsing the config.yaml to obtain the device's symmetric key and an attestation test is performed with DPS using a SAS token. The changes are primarily relegated to the dps and provisioning crates and come along with tests.

**Note:** While the feature implementation is done via this PR, it should be noted that the DPS service has not enabled provisioning devices with Edge capabilities. As a result, best effort testing has taken place and it was observed that the device that DPS registration at the destination IoT Hub took place but it did not have the Edge capability enabled. A true E2E test can take place when DPS enables this in the service.

### CODE REVIEW

File: edgelet/dps/src/error.rs [hunk 1]
```diff
@@ -40,6 +40,15 @@ pub enum ErrorKind {
 
     #[fail(display = "DPS registration failed because the DPS operation is not assigned")]
     RegisterWithAuthUnexpectedlyFailedOperationNotAssigned,
+
```
Comment (arsing) [inline edgelet/dps/src/error.rs]: Keep these in alphabetical order.

File: edgelet/dps/src/model.rs [hunk 2]
```diff
@@ -259,8 +300,11 @@ pub struct DeviceRegistrationResult {
     #[serde(rename = "deviceId", skip_serializing_if = "Option::is_none")]
     device_id: Option<String>,
     /// Enrollment status.
-    #[serde(rename = "status")]
-    status: String,
+    #[serde(rename = "status", skip_serializing_if = "Option::is_none")]
+    status: Option<String>,
+    /// Substatus for 'Assigned' devices. Possible values include - 'initialAssignment': Device has been assigned to an IoT hub for the first time, 'deviceDataMigrated': Device has been assigned to a different IoT hub and its device data was migrated from the previously assigned IoT hub. Device data was removed from the previously assigned IoT hub, 'deviceDataReset':  Device has been assigned to a different IoT hub and its device data was populated from the initial state stored in the enrollment. Device data was removed from the previously assigned IoT hub.
```
Comment (arsing) [inline edgelet/dps/src/model.rs]: Split this comment over multiple lines.

File: edgelet/iotedged/src/error.rs [hunk 3]
```diff
@@ -31,6 +31,12 @@ pub enum ErrorKind {
 
     #[fail(display = "The workload service encountered an error")]
     WorkloadService,
+
```
Comment (arsing) [inline edgelet/iotedged/src/error.rs]: Keep these in alphabetical order.

File: edgelet/iotedged/src/lib.rs [hunk 4]
```diff
@@ -299,31 +300,67 @@ impl Main {
...
-                    hyper_client,
-                    &runtime,
-                    &key_store,
-                    cfg,
-                    root_key,
-                    shutdown_signal,
-                    &crypto,
-                    tokio_runtime,
-                )?;
+                match dps.symmetric_key() {
```
Comment (arsing) [inline edgelet/iotedged/src/lib.rs]:
```rust
let (key_store, provisioning_result, root_key, runtime) = match dps.symmetric_key() {
    Some(key) => {
        info!("Staring provisioning edge device via symmetric key...");
        dps_symmetric_key_provision(...)?
    },
    None => {
        info!("Staring provisioning edge device via TPM...");
        dps_tpm_provision(...)?
    },
};

info!("Finished provisioning edge device.");
let cfg = WorkloadData::new(...);
start_api(...)?;

**Comment by mrohera** [comment]
Could I kindly get a re-approval please?

File: edgelet/iotedged/src/lib.rs [hunk 5]
```diff
@@ -614,7 +638,74 @@ fn manual_provision(
...
+    memory_hsm
+        .activate_identity_key(KeyIdentity::Device, "primary".to_string(), key_bytes)
+        .context(ErrorKind::ActivateSymmetricKey)?;
+
+    let dps = DpsSymmetricKeyProvisioning::new(
+        hyper_client,
+        provisioning.global_endpoint().clone(),
+        provisioning.scope_id().to_string(),
+        provisioning.registration_id().to_string(),
+        "2018-11-01".to_string(),
```
Comment (darobs) [inline edgelet/iotedged/src/lib.rs]: Should this be DPS_API_VERSION?

File: edgelet/dps/src/registration.rs [hunk 6]
```diff
@@ -454,17 +517,17 @@ mod tests {
...
     use tokio;
     use url::Url;
 
-    use edgelet_core::crypto::{MemoryKey, MemoryKeyStore};
-
     #[test]
-    fn server_register_with_auth_success() {
-        let expected_uri = "https://global.azure-devices-provisioning.net/scope/registrations/reg/register?api-version=2017-11-15";
+    fn server_register_with_tpm_auth_success() {
+        let expected_uri = "https://global.azure-devices-provisioning.net/scope/registrations/reg/register?api-version=2018-11-01";
```
Comment (darobs) [inline edgelet/dps/src/registration.rs]: Can we make this 
```
let expected_uri = "https://global.azure-devices-provisioning.net/scope/registrations/reg/register?api-version=" + DPS_API_VERSION;
```
?

File: edgelet/iotedged/src/lib.rs [hunk 7]
```diff
@@ -639,7 +730,7 @@ where
         provisioning.global_endpoint().clone(),
         provisioning.scope_id().to_string(),
         provisioning.registration_id().to_string(),
-        "2017-11-15".to_string(),
+        "2018-11-01".to_string(),
```
Comment (darobs) [inline edgelet/iotedged/src/lib.rs]: DPS_API_VERSION?  Should we find this in all of it's used places?

Comment (darobs) [reply edgelet/iotedged/src/lib.rs]: Is there a reason why it shouldn't be the const value?

File: edgelet/iotedged/src/settings.rs [hunk 8]
```diff
@@ -353,6 +363,24 @@ mod tests {
...
+        let settings = Settings::<DockerConfig>::new(Some(GOOD_SETTINGS_DPS_SYM_KEY));
+        println!("{:?}", settings);
+        assert!(settings.is_ok());
+        let s = settings.unwrap();
+        match s.provisioning() {
+            Provisioning::Dps(ref dps) => {
+                assert_eq!(dps.global_endpoint().scheme(), "scheme");
+                assert_eq!(dps.global_endpoint().host_str().unwrap(), "jibba-jabba.net");
+                assert_eq!(dps.scope_id(), "i got no time for the jibba-jabba");
+                assert_eq!(dps.registration_id(), "register me foo");
```
Comment (darobs) [inline edgelet/iotedged/src/settings.rs]: I'm pretty certain Mr. T would say "register me, fool" as in "I pity the fool".

Comment (mrohera) [reply edgelet/iotedged/src/settings.rs]: Thanks I will fix the typo :). It is important that Mr. T is well represented.

**Pull request merged**